Saving hyperparameters fails if all parameters in the signature get ignored

## 🐛 Bug

<!-- A clear and concise description of the bug. -->

### To Reproduce

In the test

https://github.com/PyTorchLightning/pytorch-lightning/blob/1e36cffbca3f92855d82907370544fade1656da6/tests/models/test_hparams.py#L642-L644

Add a new case that ignores all parameters:

```
@pytest.mark.parametrize("ignore", ("arg2", ("arg2", "arg3"), ("arg1", "arg2", "arg3")))
```

Error:

```
tests/models/test_hparams.py:641 (test_ignore_args_list_hparams[ignore2])
tmpdir = local('/private/var/folders/s2/338zrsts1b17v5cj18xr8_640000gn/T/pytest-of-adrian/pytest-15/test_ignore_args_list_hparams_2')
ignore = ('arg1', 'arg2', 'arg3')

    @pytest.mark.parametrize("ignore", ("arg2", ("arg2", "arg3"), ("arg1", "arg2", "arg3")))
    def test_ignore_args_list_hparams(tmpdir, ignore):
        """Tests that args can be ignored in save_hyperparameters."""
    
        class LocalModel(BoringModel):
            def __init__(self, arg1, arg2, arg3):
                super().__init__()
                self.save_hyperparameters(ignore=ignore)
    
>       model = LocalModel(arg1=14, arg2=90, arg3=50)

test_hparams.py:651: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_hparams.py:649: in __init__
    self.save_hyperparameters(ignore=ignore)
../../pytorch_lightning/core/mixins/hparams_mixin.py:107: in save_hyperparameters
    save_hyperparameters(self, *args, ignore=ignore, frame=frame)
../../pytorch_lightning/utilities/parsing.py:255: in save_hyperparameters
    obj._hparams_initial = copy.deepcopy(obj._hparams)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = LocalModel(
  (layer): Linear(in_features=32, out_features=2, bias=True)
)
name = '_hparams'

    def __getattr__(self, name: str) -> Union[Tensor, 'Module']:
        if '_parameters' in self.__dict__:
            _parameters = self.__dict__['_parameters']
            if name in _parameters:
                return _parameters[name]
        if '_buffers' in self.__dict__:
            _buffers = self.__dict__['_buffers']
            if name in _buffers:
                return _buffers[name]
        if '_modules' in self.__dict__:
            modules = self.__dict__['_modules']
            if name in modules:
                return modules[name]
>       raise AttributeError("'{}' object has no attribute '{}'".format(
            type(self).__name__, name))
E       AttributeError: 'LocalModel' object has no attribute '_hparams'

```

### Expected behavior

Just make it a no-op if all parameters get ignored. 
Motivation: The code remains agnostic to changes in the signature.

### Environment

latest torch
pl master

Ask if you need more info :) 

### Additional context

<!-- Add any other context about the problem here. -->


cc @carmocca @justusschock @awaelchli @borda @ananthsub @ninginthecloud @jjenniferdai @rohitgr7

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -603,7 +603,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Configure native Deepspeed schedulers with interval='step' ([#11788](https://github.com/PyTorchLightning/pytorch-lightning/pull/11788))
 
 
-- Fixed an `AttributeError` when calling `save_hyperparameters` when no parameters need saving ([#11827](https://github.com/PyTorchLightning/pytorch-lightning/pull/11827))
+- Fixed an `AttributeError` when calling `save_hyperparameters` and no parameters need saving ([#11827](https://github.com/PyTorchLightning/pytorch-lightning/pull/11827))
 
 
 ## [1.5.9] - 2022-01-20

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -603,6 +603,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Configure native Deepspeed schedulers with interval='step' ([#11788](https://github.com/PyTorchLightning/pytorch-lightning/pull/11788))
 
 
+- Fixed an `AttributeError` when calling `save_hyperparameters` when no parameters need saving ([#11827](https://github.com/PyTorchLightning/pytorch-lightning/pull/11827))
+
+
 ## [1.5.9] - 2022-01-20
 
 ### Fixed

--- a/pytorch_lightning/utilities/parsing.py
+++ b/pytorch_lightning/utilities/parsing.py
@@ -224,7 +224,6 @@ def save_hyperparameters(
         init_args = {}
         for local_args in collect_init_args(frame, []):
             init_args.update(local_args)
-    # assert init_args, "failed to inspect the obj init"
 
     if ignore is not None:
         if isinstance(ignore, str):

--- a/pytorch_lightning/utilities/parsing.py
+++ b/pytorch_lightning/utilities/parsing.py
@@ -224,7 +224,7 @@ def save_hyperparameters(
         init_args = {}
         for local_args in collect_init_args(frame, []):
             init_args.update(local_args)
-    assert init_args, "failed to inspect the obj init"
+    # assert init_args, "failed to inspect the obj init"
 
     if ignore is not None:
         if isinstance(ignore, str):
@@ -249,8 +249,7 @@ def save_hyperparameters(
             obj._hparams_name = "kwargs"
 
     # `hparams` are expected here
-    if hp:
-        obj._set_hparams(hp)
+    obj._set_hparams(hp)
     # make deep copy so  there is not other runtime changes reflected
     obj._hparams_initial = copy.deepcopy(obj._hparams)
 

--- a/tests/models/test_hparams.py
+++ b/tests/models/test_hparams.py
@@ -286,7 +286,6 @@ if _OMEGACONF_AVAILABLE:
             super().__init__(*args, **kwargs)
             self.save_hyperparameters()
 
-
 else:
 
     class DictConfSubClassBoringModel:

--- a/tests/models/test_hparams.py
+++ b/tests/models/test_hparams.py
@@ -286,6 +286,7 @@ if _OMEGACONF_AVAILABLE:
             super().__init__(*args, **kwargs)
             self.save_hyperparameters()
 
+
 else:
 
     class DictConfSubClassBoringModel:
@@ -670,6 +671,31 @@ def test_ignore_args_list_hparams(tmpdir, ignore):
     assert model.hparams.arg1 == 14
     for arg in ignore:
         assert arg not in model.hparams
+
+
+class IgnoreAllParametersModel(BoringModel):
+    def __init__(self, arg1, arg2, arg3):
+        super().__init__()
+        self.save_hyperparameters(ignore=("arg1", "arg2", "arg3"))
+
+
+class NoParametersModel(BoringModel):
+    def __init__(self):
+        super().__init__()
+        self.save_hyperparameters()
+
+
+@pytest.mark.parametrize(
+    "model",
+    (
+        IgnoreAllParametersModel(arg1=14, arg2=90, arg3=50),
+        NoParametersModel(),
+    ),
+)
+def test_save_no_parameters(model):
+    """Test that calling save_hyperparameters works if no parameters need saving."""
+    assert model.hparams == {}
+    assert model._hparams_initial == {}
 
 
 class HparamsKwargsContainerModel(BoringModel):

